rfd/0015-2fa-management.md

```diff
@@ -0,0 +1,206 @@
+to reset the account to recover.
+
+[RFD 14](0014-session-2FA.md) adds support for 2FA challenges per-connection.
+2FA devices will become more widely used by Enterprise users, who would
+previously delegate this to their SSO provider.
+
+Teleport needs to make 2FA device support more flexible, as both a
+long-standing OSS users' request, and the new Enterprise use-case.
+
+## Details
+
+### protocols
+
+Teleport supports two 2FA protocols: OTP and U2F.
+
+There's no change to the list of supported protocols, but the implementation
+should assume that more protocols will be added (e.g. WebAuthn).
+
+For U2F, we will migrate from https://github.com/tstranex/u2f to
+https://github.com/flynn/u2f, to allow client-side CLI authentication without
+the `u2f-host` dependency.
+
+### UX
+
+For the below examples, we assume that the user already has at least one 2FA
+device enrolled. The [bootstrap](#bootstrap) section covers registration of the
+first device.
+
+#### CLI
+
+Login:
+
+```sh
+$ tsh login --login=non-sso-user
+Enter password for Teleport user non-sso-user: ...
+Tap your security key... <tap>
+
+$ tsh login --login=non-sso-user --2fa=otp
+Enter password for Teleport user non-sso-user: ...
+Enter OTP code: ...
+
+$ tsh login --login=sso-user --auth=oidc
+# SSO page opens
+# no 2FA prompt from Teleport
+```
+
+2FA management:
+
+```sh
+$ tsh 2fa ls
+2FA deivice name   Type   Last used
+----------------   ----   -------------------------------
+android OTP        OTP    Tue 15 Dec 2020 01:29:42 PM PST
+yubikey            U2F    Wed 16 Dec 2020 02:00:13 PM PST
+
+$ tsh 2fa add
+Adding a new 2FA device.
+Choose device type (1 - OTP, 2 - U2F): 2
+Enter device name: solokey
+Tap any *registered* security key... <tap>
+Tap your *new* security key... <tap>
+2FA device "solokey" added.
+
+$ tsh 2fa ls
+2FA deivice name   Type   Last used
+----------------   ----   -------------------------------
+android OTP        OTP    Tue 15 Dec 2020 01:29:42 PM PST
+yubikey            U2F    Wed 16 Dec 2020 02:00:13 PM PST
+solokey            U2F    Wed 16 Dec 2020 02:05:46 PM PST
+
+$ tsh 2fa rm yubikey
+Tap any *registered* security key... <tap>
+2FA device "yubikey" removed.
+
+$ tsh 2fa rm "android OTP"
+Tap any *registered* security key... <tap>
+2FA device "android OTP" removed.
+
+$ tsh 2fa ls
+2FA deivice name   Type   Last used
+----------------   ----   -------------------------------
+solokey            U2F    Wed 16 Dec 2020 02:06:46 PM PST
+
+# If 2FA is optional:
+$ tsh 2fa rm solokey
+You are about to remove the only remaining 2FA device.
+This will disable 2FA during login.
+Are you sure? (y/N): N
+
+# If 2FA is required:
+$ tsh 2fa rm solokey
+Can't remove the only remaining 2FA device.
+Please add a replacement 2FA device first using "tsh 2fa add".
+```
+
+#### web UI
+
+Web UI management of 2FA devices should be logically similar to the CLI:
+- a page to see all enrolled devices
+- a button to enroll a new device
```

Comment:
Are we planning on using WebAuthn? It has a range of built in UX for registering tokens per supported browser https://caniuse.com/?search=webauthn

AWS recently added support and it has a few good points of reference. https://aws.amazon.com/blogs/aws/multi-factor-authentication-with-webauthn-for-aws-sso/

Reply:
U2F UX in the browser is basically identical to WebAuthn - you get a popup prompting for a tap.
It's not that clunky version where you have a textbox and the U2F token generates a bunch of random key presses.

But yes, at some point we can add WebAuthn support too.

Reply:
Thanks for the clarification, I guess It'll be same as our current U2F setup.. which could be improved but works.

Reply:
What are the issues with our current U2F setup (other than being browser-only)?
Maybe I can address those along the way.

Reply:
The setup is currently a bit clunky, but once setup it's pretty smooth. I'll set it up again and find the rough edges.

Reply:
WebAuthn is a part of FIDO2, the successor to U2F.  I believe it is more compatible on mobile as well (I noticed mobile NFC U2F doesn't currently work this week, at least on Android/Chrome).

Reply:
AFAIK, the user interaction would be basically the same for U2F and WebAuthn.

The only difference is token support when using a browser (e.g. SSH via the web UI).
On the CLI (`kubectl` or `tsh ssh`), we'll only support USB devices for now.